soc/st/stm32/stm32wbax/power.c

```diff
@@ -97,88 +104,130 @@ static void set_mode_suspend_to_ram(void)
 
+	/* Erratum 2.2.15:
+	 * Disabling ICACHE is required before entering stop mode
+	 */
+	sys_cache_instr_disable();
+	/* If stop mode is greater than 1, manage flash latency and RAMCFG */
+	if (substate_id > 1) {
+		flash_latency_backup = __HAL_FLASH_GET_LATENCY();
+		if (flash_latency_backup < FLASH_LATENCY_1) {
+			/* Set flash latency to 1, since system will restart from HSI16 */
+			__HAL_FLASH_SET_LATENCY(FLASH_LATENCY_1);
+			while (__HAL_FLASH_GET_LATENCY() != FLASH_LATENCY_1) {
+			}
+		}
+		ram_waitstates_backup = READ_BIT(RAMCFG_SRAM1->CR, RAMCFG_CR_WSC);
+		MODIFY_REG(RAMCFG_SRAM1->CR, RAMCFG_CR_WSC, RAMCFG_WAITSTATE_1);
+		MODIFY_REG(RAMCFG_SRAM2->CR, RAMCFG_CR_WSC, RAMCFG_WAITSTATE_1);
+	}
+	switch (substate_id) {
+	case 1:
+		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP0);
 		break;
-#if defined(CONFIG_PM_S2RAM)
-	case PM_STATE_SUSPEND_TO_RAM:
-		set_mode_suspend_to_ram();
+	case 2:
+		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP1);
 		break;
-#endif
 	default:
-		LOG_DBG("Unsupported power state %u", state);
+		LOG_DBG("Unsupported power state substate-id %u", substate_id);
 		return;
 	}
+
+	enter_low_power_mode();
 }
 
-/* Handle SOC specific activity after Low Power Mode Exit */
-void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+static void set_mode_stop_exit(uint8_t substate_id)
 {
-#ifdef CONFIG_BT_STM32WBA
-	if (LL_PWR_IsActiveFlag_STOP() == 1U) {
-		scm_setup();
+	if (LL_PWR_IsActiveFlag_STOP() || !HSE_ON) {
+		/* Reconfigure the clock */
+		stm32_clock_control_init(NULL);
 	} else {
-		scm_setwaitstates(RUN);
+		/* Stop mode skipped */
+		if (substate_id > 1) {
+			__HAL_FLASH_SET_LATENCY(flash_latency_backup);
+			while (__HAL_FLASH_GET_LATENCY() != flash_latency_backup) {
+			}
+			MODIFY_REG(RAMCFG_SRAM1->CR, RAMCFG_CR_WSC, ram_waitstates_backup);
+			MODIFY_REG(RAMCFG_SRAM2->CR, RAMCFG_CR_WSC, ram_waitstates_backup);
+		}
 	}
-#endif
 
+	/* Erratum 2.2.15:
+	 * Enable ICACHE when exiting stop mode
+	 */
+	sys_cache_instr_enable();
+
+	LL_LPM_DisableSleepOnExit();
+	LL_LPM_EnableSleep();
+}
+
+/* Invoke Low Power/System Off specific Tasks */
+void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
 	switch (state) {
 	case PM_STATE_SUSPEND_TO_IDLE:
-		if (substate_id <= 2) {
-			/* Erratum 2.2.15:
-			 * Enable ICACHE when exiting stop mode
-			 */
-			sys_cache_instr_enable();
-
-			LL_LPM_DisableSleepOnExit();
-			LL_LPM_EnableSleep();
-		} else {
-			LOG_DBG("Unsupported power substate-id %u",
-							substate_id);
-		}
+		set_mode_stop_enter(substate_id);
+		/* Resume here */
+		set_mode_stop_exit(substate_id);
 		break;
-	case PM_STATE_SUSPEND_TO_RAM:
 #if defined(CONFIG_PM_S2RAM)
-		stm32wba_init();
-		stm32_power_init();
-
-		LL_LPM_DisableSleepOnExit();
-		LL_LPM_EnableSleep();
-#else
-		LOG_DBG("Suspend to RAM needs CONFIG_PM_S2RAM to be enabled");
-#endif
+	case PM_STATE_SUSPEND_TO_RAM:
+		set_mode_suspend_to_ram_enter();
+		/* Resume here */
+		set_mode_suspend_to_ram_exit();
```

Comment:
...to check the return value here, and skip resume path if not needed:
```suggestion
		if (set_mode_suspend_to_ram_enter() >= 0) {
		   /* SoC entered in STANDBY mode successfully.
		    * Execution resumes here after wakeup.
		    */
		   set_mode_suspend_to_ram_exit();
		}
```
(this is more of an optimization than anything else)

Reply:
This optimization has not given the expected result. So I prefer to not go ahead with your suggestion.